projects/Mallard/src/hooks/use-net-info.tsx

```diff
@@ -1,22 +1,191 @@
+ * explicitly checking the netinfo status, we can override it in __DEV__
+ *
+ * Currently aeroplane mode doesn't work in the Simulator, and switching on
+ * aeroplane mode while debugging on device ends up disconnecting you from
+ * the debugger and throwing an error in the app
+ *
+ * Potentially we can wrap our fetches in checks to this status in future
+ * if it would be helpful to both not fetch when we're offline and again,
+ * mock an offline status.
+ *
+ * 2. The original `useNetInfo` hook seemed to exhibit a re-render quickly
+ * after the first mount. Now that we store the info in a state container
+ * we can immediately return the current value without having to return
+ * unknown while waiting for a promise to resolve
+ **/
+
+const { NetInfoStateType } = NetInfo
 
 const StableNetInfoContext = createContext<NetInfoState>({
     type: NetInfoStateType.unknown,
     isConnected: false,
     details: null,
 })
 
-const NetInfoProvider = ({ children }: { children: React.ReactNode }) => (
-    <StableNetInfoContext.Provider value={originalUseNetInfo()}>
-        {children}
-    </StableNetInfoContext.Provider>
+const offlineState = {
+    type: NetInfoStateType.none,
+    isConnected: false,
+    details: null,
+} as const
+
+const devToggleStyles = StyleSheet.create({
+    bg: {
+        backgroundColor: 'black',
+        padding: 8,
+        borderRadius: 999,
+        position: 'absolute',
+        bottom: 20,
+        left: 20,
+        zIndex: 999999999,
+    },
+    text: {
+        color: 'white',
+    },
+})
+
+export class NetInfoStateContainer {
+    private firstStatePromise: Promise<NetInfoState> | null
+    private actualState: NetInfoState
+    private subscribers: ((state: NetInfoState) => void)[] = []
+    private forceOffline = false
+
+    constructor(register: (fn: (state: NetInfoState) => void) => () => void) {
+        this.actualState = {
+            type: NetInfoStateType.unknown,
+            isConnected: false,
+            details: null,
+        }
+        /**
+         * this will wait for the first state and then clear itself,
+         * in order to return an accurate value for fetch for the first time
+         *
+         * it also registers the general update handler
+         */
+        this.firstStatePromise = new Promise(res =>
+            register(state => {
+                if (this.firstStatePromise) {
+                    res(state)
+                }
+                this.onInternalStateChanged(state, this.forceOffline)
+                this.firstStatePromise = null
+            }),
+        )
+    }
+
+    get state() {
+        return this.forceOffline ? offlineState : this.actualState
+    }
+
+    get isForcedOffline() {
+        return this.forceOffline
+    }
+
+    subscribe(fn: (state: NetInfoState) => void) {
+        this.subscribers.push(fn)
+        return () => {
+            this.subscribers = this.subscribers.filter(sub => sub !== fn)
+        }
+    }
```

Comment:
we're kind of reimplementing Apollo here with a subscription mechanism, where `useQuery` would already re-render on cache changes 🤔 but it's alright for now I guess